pkg/networkserver/grpc_ns_test.go

```diff
@@ -0,0 +1,98 @@
+// Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package networkserver_test
+
+import (
+	"testing"
+
+	"github.com/smartystreets/assertions"
+	"go.thethings.network/lorawan-stack/pkg/component"
+	. "go.thethings.network/lorawan-stack/pkg/networkserver"
+	"go.thethings.network/lorawan-stack/pkg/types"
+	"go.thethings.network/lorawan-stack/pkg/util/test"
+	"go.thethings.network/lorawan-stack/pkg/util/test/assertions/should"
+)
+
+func TestGenerateDevAddrAssignment(t *testing.T) {
+	a := assertions.New(t)
+
+	// Use DevAddr prefix from NetID.
+	{
+		ns := test.Must(New(
+			component.MustNew(test.GetLogger(t), &component.Config{}),
+			&Config{
+				NetID:               types.NetID{0x00, 0x00, 0x13},
+				DeduplicationWindow: 42,
+				CooldownWindow:      42,
+				DownlinkTasks: &MockDownlinkTaskQueue{
+					PopFunc: DownlinkTaskPopBlockFunc,
+				},
+			})).(*NetworkServer)
+
+		devAddr, _ := ns.GenerateDevAddrAssignment(test.Context(), nil)
+		a.So(devAddr.DevAddr.HasPrefix(types.DevAddrPrefix{
+			DevAddr: types.DevAddr{0x26, 0, 0, 0},
+			Length:  7,
+		}), should.BeTrue)
+	}
+
+	// Configured DevAddr prefixes.
+	{
+		testPrefix1 := types.DevAddrPrefix{
+			DevAddr: types.DevAddr{0x26, 0x01, 0x00, 0x00},
+			Length:  16,
+		}
+		testPrefix2 := types.DevAddrPrefix{
+			DevAddr: types.DevAddr{0x26, 0xff, 0x01, 0x00},
+			Length:  24,
+		}
+		testPrefix3 := types.DevAddrPrefix{
+			DevAddr: types.DevAddr{0x27, 0x00, 0x00, 0x00},
+			Length:  8,
+		}
+		ns := test.Must(New(
+			component.MustNew(test.GetLogger(t), &component.Config{}),
+			&Config{
+				NetID: types.NetID{0x00, 0x00, 0x13},
+				DevAddrPrefixes: []types.DevAddrPrefix{
+					testPrefix1,
+					testPrefix2,
+					testPrefix3,
+				},
+				DeduplicationWindow: 42,
+				CooldownWindow:      42,
+				DownlinkTasks: &MockDownlinkTaskQueue{
+					PopFunc: DownlinkTaskPopBlockFunc,
+				},
+			})).(*NetworkServer)
+
+		seen := map[types.DevAddrPrefix]int{}
+		for i := 0; i < 100; i++ {
+			devAddr, _ := ns.GenerateDevAddrAssignment(test.Context(), nil)
+			if devAddr.DevAddr.HasPrefix(testPrefix1) {
+				seen[testPrefix1]++
+			}
+			if devAddr.DevAddr.HasPrefix(testPrefix2) {
+				seen[testPrefix2]++
+			}
+			if devAddr.DevAddr.HasPrefix(testPrefix3) {
+				seen[testPrefix3]++
+			}
+		}
+		a.So(seen[testPrefix1], should.BeGreaterThan, 0)
```

Comment:
This is still non-deterministic, drop the `seen` and assert the prefix in the loop per each generated DevAddr.(i.e. assert that DevAddr has one of the 3 prefixes specified)

Reply:
This comes from my suggested code where we test the generation itself. I agree that we should assert that, but _also_ that all prefixes are present. With N of 100 that, statistically, must be the case if the RNG isn’t broken. I’d like to test that. We can increase N though, but all prefixes must be there. M

Reply:
So what would be the best way to resolve that, if we talk about assertion AND check if all of the prefixes present? add assertion without dropping the `seen`?

Reply:
Yeah, or sum the seen and check if that number matches the number of generated, or fatal the test if the prefix doesn't match in the loop

Reply:
But we don't need to test the RNG here, right?
Why not just do this on `L84`:
```go
hasOneOfPrefixes := func(devAddr types.DevAddr, prefixes ...string) bool {
	for _, p := range prefixes {
		if devAddr.HasPrefix(p) {
			return true
		}
	}
	return false
}
a.So(hasOneOfPrefixes(devAddr.DevAddr, testPrefix1, testPrefix2, testPrefix3), should.BeTrue)
```
It's obvious what happens here and it's easy to debug.

Reply:
>But we don't need to test the RNG here, right?

We need to test if all the configured prefixes are considered. That is _complementary_ to testing whether all DevAddrs fit in one of the configured prefixes.

Reply:
If you want to test that NS actually uses the RNG, then you can leave the `seen` assertions as is, but you do need to assert that devAddr generated is actually valid

Reply:
Yes, that is the long version of "complementary" 😉